File: src/0071.simplify-path/main.cpp
// Copyright (c) 2024 Xu Shaohua <shaohua@biofan.org>. All rights reserved.
// Use of this source is governed by General Public License that can be found
// in the LICENSE file.

#include <cassert>

#include <sstream>
#include <string>
#include <vector>

std::string simplify_path(std::string path) {
  // 创建栈, 用于存放每个目录名.
  std::vector<std::string> stack;
  std::stringstream ss(path);
  std::string folder_name;
  while (std::getline(ss, folder_name, '/')) {
    if (folder_name.empty()) {
      // 忽略空白
    } else if (folder_name == ".") {
      // 忽略当前目录
    } else if (folder_name == "..") {
      // 返回上个目录
      if (!stack.empty()) {
        stack.pop_back();
      }
    } else {
      stack.emplace_back(folder_name);
    }
  }

  std::string real_path;
  if (stack.empty()) {
    // 目录以 "/" 开头
    real_path += "/";
  } else {
    for (const auto& folder_name : stack) {
      real_path = real_path + "/" + folder_name;
    }
  }

  return real_path;
}

void check_solution() {
  const auto s1 = simplify_path("/home/");
  assert(s1 == "/home");

  const auto s2 = simplify_path("/home//foo/");
  assert(s2 == "/home/foo");

  const auto s3 = simplify_path("/home/user/Documents/../Pictures");
  assert(s3 == "/home/user/Pictures");

  const auto s4 = simplify_path("/../");
  assert(s4 == "/");

  const auto s5 = simplify_path("/.../a/../b/c/../d/./");
  assert(s5 == "/.../b/d");
}

int main() {
  check_solution();
  return 0;
}

File: src/0071.simplify-path/src/main.rs
// Copyright (c) 2024 Xu Shaohua <shaohua@biofan.org>. All rights reserved.
// Use of this source is governed by General Public License that can be found
// in the LICENSE file.

// Stack
pub fn simplify_path1(path: String) -> String {
    let mut stack: Vec<String> = Vec::new();
    for part in path.split('/') {
        match part {
            // 忽略空白
            "" => (),
            // 忽略当前目录
            "." => (),
            // 返回上个目录
            ".." => {
                let _ = stack.pop();
            }
            part => stack.push(part.to_owned()),
        }
    }
    // 根目录
    let mut result = "/".to_owned();
    result.push_str(&stack.join("/"));
    result
}

// Stack
// 优化字符串操作
pub fn simplify_path2(path: String) -> String {
    let mut stack: Vec<&str> = Vec::with_capacity(path.len());
    for part in path.split('/') {
        match part {
            // 忽略空白
            "" => (),
            // 忽略当前目录
            "." => (),
            // 返回上个目录
            ".." => {
                let _ = stack.pop();
            }
            part => stack.push(part),
        }
    }
    // 目录以 "/" 开头
    format!("/{}", stack.join("/"))
}

pub type SolutionFn = fn(String) -> String;

fn check_solution(func: SolutionFn) {
    let path = "/home/".to_owned();
    assert_eq!(&func(path), "/home");

    let path = "/home//foo/".to_owned();
    assert_eq!(&func(path), "/home/foo");

    let path = "/home/user/Documents/../Pictures".to_owned();
    assert_eq!(&func(path), "/home/user/Pictures");

    let path = "/../".to_owned();
    assert_eq!(&func(path), "/");

    let path = "/.../a/../b/c/../d/./".to_owned();
    assert_eq!(&func(path), "/.../b/d");
}

fn main() {
    check_solution(simplify_path1);
    check_solution(simplify_path2);
}

#[cfg(test)]
mod tests {
    use super::{check_solution, simplify_path1, simplify_path2};

    #[test]
    fn test_simplify_path1() {
        check_solution(simplify_path1);
    }

    #[test]
    fn test_simplify_path2() {
        check_solution(simplify_path2);
    }
}
